tensorflow/compiler/xla/service/conditional_to_select.cc

```diff
@@ -0,0 +1,102 @@
+/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/compiler/xla/service/conditional_to_select.h"
+
+#include "tensorflow/compiler/xla/service/call_graph.h"
+#include "tensorflow/compiler/xla/service/call_inliner.h"
+#include "tensorflow/compiler/xla/service/hlo_computation.h"
+#include "tensorflow/compiler/xla/service/hlo_instruction.h"
+#include "tensorflow/compiler/xla/service/hlo_opcode.h"
+#include "tensorflow/compiler/xla/status_macros.h"
+#include "tensorflow/compiler/xla/types.h"
+#include "tensorflow/core/lib/core/errors.h"
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/platform/logging.h"
+
+namespace xla {
+
+static StatusOr<bool> DoConditionalToSelect(HloInstruction* conditional) {
+  // Only allow conditional to select if the called computations
+  // do not have side effects.
+  if (conditional->true_computation()->HasSideEffect() ||
+      conditional->false_computation()->HasSideEffect()) {
+    VLOG(1) << "Not transforming conditional; branches have side effects:"
+            << conditional->ToString();
+    return false;
+  }
+
+  auto computation = conditional->parent();
+
+  // Create new instructions
+  HloInstruction* if_call_op =
+      computation->AddInstruction(HloInstruction::CreateCall(
+          conditional->shape(), {conditional->mutable_operand(1)},
+          conditional->true_computation()));
+  HloInstruction* else_call_op =
+      computation->AddInstruction(HloInstruction::CreateCall(
+          conditional->shape(), {conditional->mutable_operand(2)},
+          conditional->false_computation()));
+  HloInstruction* condition = conditional->mutable_operand(0);
+  // If the output is not scalar, we need to broadcast the condition
+  // to match the contract of kSelect. For tuples, we use kTupleSelect
+  // which expects the condition to be a scalar.
+  if (!ShapeUtil::IsScalar(conditional->shape()) &&
+      !ShapeUtil::IsTuple(conditional->shape())) {
+    condition = computation->AddInstruction(HloInstruction::CreateBroadcast(
```

Comment:
This can be a new utility in tensorflow/compiler/xla/service/hlo_creation_utils.h

Reply:
Making the existing switch between select and triple-select sgtm.